lib/modules/manager/devcontainer/extract.spec.ts

```diff
@@ -0,0 +1,341 @@
+import { codeBlock } from 'common-tags';
+import * as dockerfileExtract from '../dockerfile/extract';
+import { extractPackageFile } from '.';
+
+describe('modules/manager/devcontainer/extract', () => {
+  describe('extractPackageFile()', () => {
+    afterEach(() => {
+      jest.restoreAllMocks();
+    });
+
+    it('returns null when the dev container JSON file is empty', () => {
+      // Arrange
+      const content = '';
+      const packageFile = '';
+      const extractConfig = {};
+      // Act
+      const result = extractPackageFile(content, packageFile, extractConfig);
+
+      // Assert
+      expect(result).toBeNull();
+    });
+
+    it('returns null when the dev container JSON file contents are malformed', () => {
+      // Arrange
+      const content = 'malformed json}}}';
+      const packageFile = '';
+      const extractConfig = {};
+      // Act
+      const result = extractPackageFile(content, packageFile, extractConfig);
+
+      // Assert
+      expect(result).toBeNull();
+    });
+
+    it('returns feature image deps when only the features property is defined in dev container JSON file', () => {
+      // Arrange
+      const content = codeBlock(`
+      {
+        "features": {
+          "devcontainer.registry.renovate.com/test/features/first:1.2.3": {},
+          "devcontainer.registry.renovate.com/test/features/second:4.5.6": {}
+        }
+      }`);
+      const extractConfig = {};
+      // Act
+      const result = extractPackageFile(
+        content,
+        'devcontainer.json',
+        extractConfig,
+      );
+
+      // Assert
+      expect(result?.deps).toMatchInlineSnapshot(`
+      [
+        {
+          "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
+          "currentDigest": undefined,
+          "currentValue": "1.2.3",
+          "datasource": "docker",
+          "depName": "devcontainer.registry.renovate.com/test/features/first",
+          "replaceString": "devcontainer.registry.renovate.com/test/features/first:1.2.3",
+        },
+        {
+          "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
+          "currentDigest": undefined,
+          "currentValue": "4.5.6",
+          "datasource": "docker",
+          "depName": "devcontainer.registry.renovate.com/test/features/second",
+          "replaceString": "devcontainer.registry.renovate.com/test/features/second:4.5.6",
+        },
+      ]
+      `);
+    });
```

Comment:
```suggestion
      expect(result).toEqual({
      deps:
      [
        {
          "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
          "currentDigest": undefined,
          "currentValue": "1.2.3",
          "datasource": "docker",
          "depName": "devcontainer.registry.renovate.com/test/features/first",
          "replaceString": "devcontainer.registry.renovate.com/test/features/first:1.2.3",
        },
        {
          "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
          "currentDigest": undefined,
          "currentValue": "4.5.6",
          "datasource": "docker",
          "depName": "devcontainer.registry.renovate.com/test/features/second",
          "replaceString": "devcontainer.registry.renovate.com/test/features/second:4.5.6",
        },
      ]
      });
    });
```

needs prettier formatting